src/Microsoft.Extensions.PlatformAbstractions/Native/PlatformApis.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.PlatformAbstractions.Native
+{
+    internal static class PlatformApis
+    {
+        private class DistroInfo
+        {
+            public string Id;
+            public string VersionId;
+        }
+
+        private static readonly Lazy<Platform> _platform = new Lazy<Platform>(DetermineOSPlatform);
+        private static readonly Lazy<DistroInfo> _distroInfo = new Lazy<DistroInfo>(LoadDistroInfo);
+
+        public static string GetOSName()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return "Windows";
+                case Platform.Linux:
+                    return GetDistroId();
+                case Platform.Darwin:
+                    return "Mac OS X";
+                default:
+                    return "Unknown";
+            }
+        }
+
+        public static string GetOSVersion()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return NativeMethods.Windows.RtlGetVersion();
+                case Platform.Linux:
+                    return GetDistroVersionId();
+                case Platform.Darwin:
+                    return GetDarwinVersion();
+                default:
+                    return string.Empty;
+            }
+        }
+
+        private static string GetDarwinVersion()
+        {
+            var kernelRelease = NativeMethods.Darwin.GetKernelRelease();
+            var version = Version.Parse(kernelRelease);
+            if (version.Major < 5)
+            {
+                // 10.0 covers all versions prior to Darwin 5
+                return "10.0";
+            }
+            else
+            {
+                // Mac OS X 10.1 mapped to Darwin 5.x, and the mapping continues that way
+                // So just subtract for from the Darwin version.
+                // https://en.wikipedia.org/wiki/Darwin_%28operating_system%29
+                return $"10.{version.Major - 4}";
+            }
+        }
+
+        public static Platform GetOSPlatform()
+        {
+            return _platform.Value;
+        }
+
+        private static string GetDistroId()
+        {
+            return _distroInfo.Value?.Id;
+        }
+
+        private static string GetDistroVersionId()
+        {
+            return _distroInfo.Value?.VersionId;
+        }
+
+        private static DistroInfo LoadDistroInfo()
+        {
+            try
+            {
+                var lines = File.ReadAllLines("/etc/os-release");
```

Comment:
In general I agree but I don't agree with "at all costs". In this particular case - when is it a valid case to not have `/etc/os-release` and if this happens you are already on slippery slope - likely network won't work, you don't have the right libicu etc. In this case the exception would just signal the very first bump. In addition all people who are actually running on supported distros won't be punished by additional I/O that is required to avoid a first chance exception on unsupported OSes.